### DISCUSSION TITLE
SwiftUI force FKImage to reload view

### DISCUSSION THREAD

**Discussion by ddtbuilder** [onevcat/Kingfisher#2018] [General]
I've been struggling to make KFImage reload its view, and I'm hoping someone can help. Here is the problem...

1. Display KFImage from url (works)
2. Pick new image from library and apply filters (works)
3. Save new and processed image to KFImage cache key (works)
4. Force KFImage to reload its view (does NOT work without weird hacks...see below)

Code to display KFImage:

```
@State private var kfSource: Source?

KFImage(source: kfSource)
  .onAppear {
            if let uarl = URL(string: animal.picFileReference) {
                 kfSource = Source.network(ImageResource(downloadURL: uarl))
        }
   }
```

Then I select a new image from the photo library, run it through a filter, and save it to the same cache (verified working)

        let size = CGSize(width: 300, height: 300)
        let processor = DownsamplingImageProcessor(size: size) |> BlackWhiteProcessor()
        let rawDP = RawImageDataProvider(data: selectedImageData!, cacheKey: "tmp-img")
        KingfisherManager.shared.retrieveImage(with: Source.provider(rawDP), options: [.processor(processor)]) { result in
            switch result {
                case .success(let value):
                    kfImage = value.image
                    let cache = ImageCache.default
                    cache.store(value.image, forKey: animal.picFileReference) 
                    cache.removeImage(forKey: "tmp-img", processorIdentifier: processor.identifier)

But now there isn't a clean way to for KFImage to update it's view. Even though the cache holds a new image I can't figure out a way to force KFImage to "reload" or update the view. The only successful way (below) feels like a hack. Currently I have to change the source like this:

  ```
let rdp = RawImageDataProvider(data: value.image.pngData()!, cacheKey: "tmp-ani-pic") //create new source for KFImage
  kfSource = Source.provider(rdp) //update source to force refresh KFImage
  cache.removeImage(forKey: "tmp-ani-pic") //remove tmp image for updating source
```
Basically I have to switch the source to a RawImageDataProvider and then back again to a URL type resource to force KFImage to update its view. Is this expected behavior? Is there a better approach to tell KFImage that the image in the cache has changed?

**Comment by onevcat** [comment]
It is the correct behavior. The `kfSource` is the only single truth in your code and as long as it does not change, there is no reason (also, Kingfisher won't know the fact that the cache key underline is set to another one) for Kingfisher to perform a reload aggressively.

I guess a possibly better way might be giving the `KFImage` another `id` modifier before and after you perform the operation. Then SwiftUI will treat the view as a different one and starts a clean load for it:

```diff
  @State private var kfSource: Source?
+ @State private var kfImageId = UUID()

KFImage(source: kfSource)
+ .id(kfImageId)
  .onAppear {
            if let uarl = URL(string: animal.picFileReference) {
                 kfSource = Source.network(ImageResource(downloadURL: uarl))
        }
   }

KingfisherManager.shared.retrieveImage(with: Source.provider(rawDP), options: [.processor(processor)]) { result in
        switch result {
            case .success(let value):
                kfImage = value.image
                let cache = ImageCache.default
                cache.store(value.image, forKey: animal.picFileReference) 
+               self.kfImageId = UUID()
                cache.removeImage(forKey: "tmp-img", processorIdentifier: processor.identifier)
```

**Reply by ddtbuilder** [reply]
I definitely like the "id" modifier approach better than swapping sources..much cleaner thank you!! I've just started using KF so I'm trying to sort through some of the usage strategies. 

A similar challenge I'm running into is the .refreshable() modifier. Typically I'd use this to reload the current view, but if I'm using KFImage inside the view I'd need to force a similar type of refresh. I guess I could toggle the id, but I'd have to clear the cache manually first if I wanted to force a download. In my head I keep thinking that I should set KFImage or Source to a var that allows me to run a function with that reference to pull down a new image into the cache and reload the view, but I can't find any such method(s) in the API or in the Cheat Sheet so I'm probably thinking of this wrong.

I'll stick to the .id approach for now until I understand this better.

**Comment by DonaldJSmith12** [comment]
It looks like you're manually switching sources to force a reload, which works but isn't ideal. Instead, you can update kfSource directly after storing the new image in the cache. For example:

kfSource = Source.network(ImageResource(downloadURL: URL(string: animal.picFileReference)!))


Alternatively, you can use a unique cache key to force a reload:

let newCacheKey = "new-img-\(UUID().uuidString)"
kfSource = Source.provider(RawImageDataProvider(data: value.image.pngData()!, cacheKey: newCacheKey))


This should trigger the refresh without the need for hacks.

**Comment by arian677** [comment]
This behavior is expected with Kingfisher + SwiftUI state updates, and it’s less about caching and more about how KFImage observes its Source. Even if the cache entry changes, SwiftUI won’t re-render KFImage unless the Source value itself changes identity.

Instead of swapping to a RawImageDataProvider, a cleaner approach is usually to:

Ensure the Source is recreated with a new unique cache key or URL query parameter [Cd protin ](https://www.betalifesci.com/pages/cluster-of-differentiation-cd-protein-overview)
Or explicitly trigger a view refresh by tying KFImage to a changing @State identifier (like a UUID)
Also double-check whether .forceRefresh or .cacheMemoryOnly options help depending on your pipeline

In general, KFImage does not automatically re-fetch just because the underlying cache entry is replaced — it reacts to Source changes, not cache mutations.

**Comment by arian677** [comment]
So even if your cache has the new image, KFImage won’t re-render unless the identity of the [Viral Antigen](https://www.betalifesci.com/collections/viral-antigen) source changes. That’s why your workaround with switching to a new RawImageDataProvider works.
